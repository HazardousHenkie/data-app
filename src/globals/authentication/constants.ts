import AuthenticationState, { User } from './types'

export const initialUserState: User = {
    id: '',
    name: ''
}

const initialAuthenticationState: AuthenticationState = {
    loggedIn: false,
    error: false,
    loading: false,
    user: initialUserState
}

export default initialAuthenticationState
