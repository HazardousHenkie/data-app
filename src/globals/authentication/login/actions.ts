import { action } from 'typesafe-actions'

import AuthenticationState from '../types'
import ActionTypes from './constants'

export const loginRequest = (loginToken: string) =>
    action(ActionTypes.LOGIN_REQUEST, loginToken)
export const loginSuccess = (user: AuthenticationState['user']) =>
    action(ActionTypes.LOGIN_SUCCESS, user)
export const loginError = (error: Error | boolean) =>
    action(ActionTypes.LOGIN_ERROR, error)
