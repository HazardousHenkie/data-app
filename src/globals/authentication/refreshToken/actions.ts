import { action } from 'typesafe-actions'

import AuthenticationState from '../types'
import ActionTypes from './constants'

export const getRefreshTokenRequest = (id: string) =>
    action(ActionTypes.GET_REFRESH_TOKEN_REQUEST, id)
export const getRefreshTokenSuccess = (user: AuthenticationState['user']) =>
    action(ActionTypes.GET_REFRESH_TOKEN_SUCCESS, user)
export const getRefreshTokenError = (error: Error | boolean) =>
    action(ActionTypes.GET_REFRESH_TOKEN_ERROR, error)
