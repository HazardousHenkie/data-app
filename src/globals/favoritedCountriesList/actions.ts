import { action } from 'typesafe-actions'

import { ResponseError } from 'utils/request'
import ActionTypes from './constants'

import FavoritedCountriesState from './types'

export const getFavoritedCountries = () =>
    action(ActionTypes.GET_FAVORITED_COUNTRIES)

export const setFavoritedCountries = (
    favoritedCountries: FavoritedCountriesState['countries']
) => action(ActionTypes.SET_FAVORITED_COUNTRIES, favoritedCountries)

export const getFavoritedCountriesError = (error: ResponseError) =>
    action(ActionTypes.GET_FAVORITED_COUNTRIES_ERROR, error)
