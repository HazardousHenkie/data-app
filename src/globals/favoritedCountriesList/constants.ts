import { query } from 'faunadb'
import FavoritedCountriesState from './types'

enum ActionTypes {
    GET_FAVORITED_COUNTRIES = 'components/FavoritedCountriesList/GET_FAVORITED_COUNTRIES',
    SET_FAVORITED_COUNTRIES = 'components/FavoritedCountriesList/SET_FAVORITED_COUNTRIES',
    GET_FAVORITED_COUNTRIES_ERROR = 'components/FavoritedCountriesList/GET_FAVORITED_COUNTRIES_ERROR'
}

export const initialFavoritedCountriesState: FavoritedCountriesState = {
    error: false,
    loading: false,
    countries: [
        {
            ref: {
                '@ref': {
                    collection: query.Collection('country_user'),
                    id: ''
                }
            },
            ts: 0,
            data: {
                userId: '',
                countryId: '',
                updatedAt: 0,
                createdAt: 0
            }
        }
    ]
}

export default ActionTypes
