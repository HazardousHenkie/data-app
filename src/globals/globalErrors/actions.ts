import { action } from 'typesafe-actions'

import ActionTypes from './constants'

export const getErrors = () => action(ActionTypes.GET_ERRORS)

export const setError = (error: Error) => action(ActionTypes.SET_ERROR, error)

export const removeError = () => action(ActionTypes.REMOVE_ERROR)

export const setErrors = (errors: Error[]) =>
    action(ActionTypes.SET_ERRORS, errors)
