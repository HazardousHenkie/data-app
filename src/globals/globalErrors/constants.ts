import ErrorsState from './types'

enum ActionTypes {
    GET_ERRORS = 'components/globalErrors/GET_ERRORS',
    SET_ERROR = 'components/globalErrors/SET_ERROR',
    REMOVE_ERROR = 'components/globalErrors/REMOVE_ERROR',
    SET_ERRORS = 'components/globalErrors/SET_ERRORS'
}

export const initialErrorsState: ErrorsState = {
    errors: []
}

export default ActionTypes
