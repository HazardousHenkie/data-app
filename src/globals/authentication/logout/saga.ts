import { call, put } from 'redux-saga/effects'

import requestErrorCheck from 'utils/errorCheckRequest'
import request from 'utils/request'

import ERROR_STATUS_CODES from 'utils/errorStatusCodes'
import { setError } from 'globals/globalErrors/actions'
import { logoutSuccess, logoutError } from './actions'

import authToken from '../authToken'

export default function* logoutSaga() {
    const requestURL = '/.netlify/functions/googleLogout'

    try {
        yield call(request, requestURL, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${authToken.token}`
            }
        })

        localStorage.removeItem('userId')
        yield put(logoutSuccess())
    } catch (error) {
        if (error.response.status === ERROR_STATUS_CODES.UNAUTHORIZED) {
            yield requestErrorCheck(error)
        }

        yield put(logoutError(error))
        yield put(setError(error))
    }
}
