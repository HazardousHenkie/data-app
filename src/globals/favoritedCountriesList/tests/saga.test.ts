import { put } from 'redux-saga/effects'

import { getRefreshTokenRequest } from 'globals/authentication/refreshToken/actions'
import { logoutRequest } from 'globals/authentication/logout/actions'
import { ResponseError } from 'utils/request'
import { setError } from 'globals/globalErrors/actions'

import { setFavoritedCountries, getFavoritedCountriesError } from '../actions'

import { initialFavoritedCountriesState } from '../constants'

import getFavoritedCountriesDataSaga from '../saga'

describe('getFavoritedCountriesDataSaga Saga', () => {
    let favoritedCountriesGenerator: Generator
    let callDescriptor: Response

    beforeEach(() => {
        favoritedCountriesGenerator = getFavoritedCountriesDataSaga()

        callDescriptor = favoritedCountriesGenerator.next().value
        expect(callDescriptor).toMatchSnapshot()
    })

    it('should dispatch the setFavoritedCountries action if call was successfull', () => {
        const response = { data: initialFavoritedCountriesState.countries }

        const putDescriptor = favoritedCountriesGenerator.next(response).value

        expect(putDescriptor).toEqual(
            // eslint-disable-next-line redux-saga/no-unhandled-errors
            put(setFavoritedCountries(response.data))
        )
    })

    it('should call refreshtoken if localstorage(userid) is set and there is responserror with a 401 error', () => {
        localStorage.setItem('userId', 'userId')

        const responseError = Response.error()
        // @ts-ignore
        responseError.status = 401

        const response = new ResponseError(responseError, 'Some error')

        const putDescriptor = favoritedCountriesGenerator.throw(response).value

        expect(putDescriptor).toEqual(
            // eslint-disable-next-line redux-saga/no-unhandled-errors
            put(
                getRefreshTokenRequest(localStorage.getItem('userId') as string)
            )
        )
    })

    it('should call the logoutRequest if the response errors', () => {
        const response = new ResponseError(callDescriptor, 'Some error')

        const putDescriptor = favoritedCountriesGenerator.throw(response).value

        // eslint-disable-next-line redux-saga/no-unhandled-errors
        expect(putDescriptor).toEqual(put(logoutRequest()))
    })

    it('should call the setError and logoutError action if the response errors', () => {
        const response = new ResponseError(callDescriptor, 'Some error')

        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
        favoritedCountriesGenerator.throw(response).value
        const putDescriptorSecondError = favoritedCountriesGenerator.next(
            response
        ).value

        expect(putDescriptorSecondError).toEqual(
            // eslint-disable-next-line redux-saga/no-unhandled-errors
            put(getFavoritedCountriesError(response))
        )

        const putDescriptorThirdError = favoritedCountriesGenerator.next(
            response
        ).value

        // eslint-disable-next-line redux-saga/no-unhandled-errors
        expect(putDescriptorThirdError).toEqual(put(setError(response)))
    })
})
