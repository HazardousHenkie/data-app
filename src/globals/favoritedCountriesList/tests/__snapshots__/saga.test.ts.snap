// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFavoritedCountriesDataSaga Saga should call refreshtoken if localstorage(userid) is set and there is responserror with a 401 error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/getFavoritedCountries",
      Object {
        "headers": Object {
          "Authorization": "Bearer ",
        },
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getFavoritedCountriesDataSaga Saga should call the logoutRequest if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/getFavoritedCountries",
      Object {
        "headers": Object {
          "Authorization": "Bearer ",
        },
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getFavoritedCountriesDataSaga Saga should call the setError and logoutError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/getFavoritedCountries",
      Object {
        "headers": Object {
          "Authorization": "Bearer ",
        },
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getFavoritedCountriesDataSaga Saga should dispatch the setFavoritedCountries action if call was successfull 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/getFavoritedCountries",
      Object {
        "headers": Object {
          "Authorization": "Bearer ",
        },
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
